catroid/src/main/java/org/catrobat/catroid/web/ProjectDownloader.kt

```diff
@@ -0,0 +1,183 @@
+                return
+            }
+
+            queue.enqueue(projectName)
+            startService(projectName, context, renameProject)
+        }
+    }
+
+    @VisibleForTesting
+    fun startService(projectName: String, context: Context, renameProject: Boolean) {
+        val downloadIntent = Intent(context, ProjectDownloadService::class.java)
+        downloadIntent.putExtra(ProjectDownloadService.EXTRA_DOWNLOAD_NAME, projectName)
+        downloadIntent.putExtra(ProjectDownloadService.EXTRA_URL, url)
+        downloadIntent.putExtra(ProjectDownloadService.EXTRA_RENAME_AFTER_DOWNLOAD, renameProject)
+        downloadIntent.putExtra(ProjectDownloadService.EXTRA_RESULT_RECEIVER, Receiver(projectName, Handler()))
+
+        val statusBarNotificationManager = StatusBarNotificationManager.getInstance()
+        val id = statusBarNotificationManager.createProjectDownloadNotification(context, projectName)
+        downloadIntent.putExtra(ProjectDownloadService.EXTRA_NOTIFICATION_ID, id)
+        statusBarNotificationManager.showOrUpdateNotification(context, id, 0)
+
+        callbackWeakReference.get()?.onDownloadStarted(url)
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            context.startForegroundService(downloadIntent)
+        } else {
+            context.startService(downloadIntent)
+        }
+    }
+
+    @SuppressLint("ParcelCreator")
+    private inner class Receiver internal constructor(
+        val projectName: String,
+        handler: Handler
+    ) : ResultReceiver(handler) {
+        override fun onReceiveResult(resultCode: Int, resultData: Bundle) {
+            super.onReceiveResult(resultCode, resultData)
+            when (resultCode) {
+                UPDATE_PROGRESS_CODE -> {
+                    val progress = resultData.getLong(UPDATE_PROGRESS_EXTRA)
+                    callbackWeakReference.get()?.onDownloadProgress(progress.toInt(), url)
+                }
+                SUCCESSFUL_CODE -> {
+                    callbackWeakReference.get()?.onDownloadFinished(projectName, url)
+                    queue.finished(projectName)
+                }
+                ERROR_CODE -> queue.finished(projectName)
+            }
+        }
+    }
+
+    interface ProjectDownloadQueue {
+        fun projectNameNotAvailable(projectName: String): Boolean
+        fun alreadyInQueue(projectName: String): Boolean
+        fun enqueue(projectName: String)
+        fun finished(projectName: String)
+    }
+}
+
+object GlobalProjectDownloadQueue {
+    val queue = ProjectDownloadQueue()
+
+    class ProjectDownloadQueue : ProjectDownloader.ProjectDownloadQueue {
+        private var projectNameSet: Set<String> = Collections.synchronizedSet(HashSet<String>())
+
+        @Synchronized
+        override fun projectNameNotAvailable(projectName: String): Boolean =
+            alreadyInQueue(projectName) || File(DEFAULT_ROOT_DIRECTORY, projectName).exists()
```

Comment:
please remove this method from the download queue, theres already some functionality somewhere that provides you a list of local projects that you can check against. (without having this class know about the default root directory or files there)

Reply:
This function makes the whole thing way too slow (validate string while renaming etc..) => wrote a new function. See in ReplaceExistingProjectDialogFragment